packages/components/src/box-control/axial-input-controls.tsx

```diff
@@ -66,6 +75,21 @@ export default function AxialInputControls( {
 		}
 	};
 
+	const sliderOnChange = ( side: GroupedSide, next: string ) => {
+		const nextValues = { ...values };
+
+		if ( side === 'vertical' ) {
+			nextValues.top = next;
+			nextValues.bottom = next;
+		}
+
+		if ( side === 'horizontal' ) {
+			nextValues.left = next;
+			nextValues.right = next;
+		}
+		onChange?.( nextValues );
+	};
+
```

Comment:
Similarly to what suggested for the "all input controls" scenario, we could apply the same refactor everywhere.

<details>

<summary>It could look something like this</summary>

```diff
diff --git a/packages/components/src/box-control/all-input-control.tsx b/packages/components/src/box-control/all-input-control.tsx
index 814363b3b6..863cf13ab3 100644
--- a/packages/components/src/box-control/all-input-control.tsx
+++ b/packages/components/src/box-control/all-input-control.tsx
@@ -48,7 +48,7 @@ export default function AllInputControl( {
 		onFocus( event, { side: 'all' } );
 	};
 
-	const onValueChange = ( next?: string ) => {
+	const handleOnValueChange = ( next?: string ) => {
 		const isNumeric = next !== undefined && ! isNaN( parseFloat( next ) );
 		const nextValue = isNumeric ? next : undefined;
 		const nextValues = applyValueToSides( values, nextValue, sides );
@@ -56,16 +56,6 @@ export default function AllInputControl( {
 		onChange( nextValues );
 	};
 
-	const unitControlOnChange: UnitControlProps[ 'onChange' ] = ( next ) => {
-		onValueChange( next );
-	};
-
-	const sliderOnChange = ( next?: number ) => {
-		onValueChange(
-			next !== undefined ? [ next, parsedUnit ].join( '' ) : undefined
-		);
-	};
-
 	// Set selected unit so it can be used as fallback by unlinked controls
 	// when individual sides do not have a value containing a unit.
 	const handleOnUnitChange: UnitControlProps[ 'onUnitChange' ] = ( unit ) => {
@@ -99,7 +89,7 @@ export default function AllInputControl( {
 				id={ inputId }
 				isOnly
 				value={ parsedQuantity }
-				onChange={ unitControlOnChange }
+				onChange={ handleOnValueChange }
 				onUnitChange={ handleOnUnitChange }
 				onFocus={ handleOnFocus }
 				onHoverOn={ handleOnHoverOn }
@@ -113,7 +103,13 @@ export default function AllInputControl( {
 				aria-labelledby={ inputId }
 				hideLabelFromVision
 				label={ LABELS.all }
-				onChange={ sliderOnChange }
+				onChange={ ( next ) =>
+					handleOnValueChange(
+						next !== undefined
+							? [ next, parsedUnit ].join( '' )
+							: undefined
+					)
+				}
 				min={ 0 }
 				max={ CUSTOM_VALUE_SETTINGS[ parsedUnit ?? 'px' ]?.max ?? 10 }
 				step={
diff --git a/packages/components/src/box-control/axial-input-controls.tsx b/packages/components/src/box-control/axial-input-controls.tsx
index 9fba588b77..34720a7ef6 100644
--- a/packages/components/src/box-control/axial-input-controls.tsx
+++ b/packages/components/src/box-control/axial-input-controls.tsx
@@ -75,22 +75,7 @@ export default function AxialInputControls( {
 		}
 	};
 
-	const sliderOnChange = ( side: GroupedSide, next: string ) => {
-		const nextValues = { ...values };
-
-		if ( side === 'vertical' ) {
-			nextValues.top = next;
-			nextValues.bottom = next;
-		}
-
-		if ( side === 'horizontal' ) {
-			nextValues.left = next;
-			nextValues.right = next;
-		}
-		onChange?.( nextValues );
-	};
-
-	const createHandleOnChange = ( side: GroupedSide ) => ( next?: string ) => {
+	const handleOnValueChange = ( side: GroupedSide, next?: string ) => {
 		if ( ! onChange ) {
 			return;
 		}
@@ -161,7 +146,9 @@ export default function AxialInputControls( {
 								parsedQuantity,
 								selectedUnit ?? parsedUnit,
 							].join( '' ) }
-							onChange={ createHandleOnChange( side ) }
+							onChange={ ( newValue ) =>
+								handleOnValueChange( side, newValue )
+							}
 							onUnitChange={ createHandleOnUnitChange( side ) }
 							onFocus={ createHandleOnFocus( side ) }
 							onHoverOn={ createHandleOnHoverOn( side ) }
@@ -175,15 +162,17 @@ export default function AxialInputControls( {
 							aria-labelledby={ inputId }
 							hideLabelFromVision
 							label={ LABELS[ side ] }
-							onChange={ ( newValue ) => {
-								sliderOnChange(
+							onChange={ ( newValue ) =>
+								handleOnValueChange(
 									side,
-									[
-										newValue,
-										selectedUnit ?? parsedUnit,
-									].join( '' )
-								);
-							} }
+									newValue !== undefined
+										? [
+												newValue,
+												selectedUnit ?? parsedUnit,
+										  ].join( '' )
+										: undefined
+								)
+							}
 							min={ 0 }
 							max={
 								CUSTOM_VALUE_SETTINGS[ selectedUnit ?? 'px' ]
diff --git a/packages/components/src/box-control/input-controls.tsx b/packages/components/src/box-control/input-controls.tsx
index 82a9701959..98713c78d8 100644
--- a/packages/components/src/box-control/input-controls.tsx
+++ b/packages/components/src/box-control/input-controls.tsx
@@ -14,7 +14,7 @@ import {
 	InputWrapper,
 } from './styles/box-control-styles';
 import type { BoxControlInputControlProps, BoxControlValue } from './types';
-import type { UnitControlProps } from '../unit-control/types';
+// import type { UnitControlProps } from '../unit-control/types';
 
 const noop = () => {};
 
@@ -49,50 +49,43 @@ export default function BoxInputControls( {
 		onChange( nextValues );
 	};
 
-	const sliderOnChange = ( side: keyof BoxControlValue, next: string ) => {
+	const handleOnValueChange = (
+		side: keyof BoxControlValue,
+		next?: string,
+		extra?: { event: React.SyntheticEvent< Element, Event > }
+	) => {
 		const nextValues = { ...values };
-		nextValues[ side ] = next;
-		handleOnChange( nextValues );
-	};
-
-	const createHandleOnChange: (
-		side: keyof BoxControlValue
-	) => UnitControlProps[ 'onChange' ] =
-		( side ) =>
-		( next, { event } ) => {
-			const nextValues = { ...values };
-			const isNumeric =
-				next !== undefined && ! isNaN( parseFloat( next ) );
-			const nextValue = isNumeric ? next : undefined;
-
-			nextValues[ side ] = nextValue;
-
-			/**
-			 * Supports changing pair sides. For example, holding the ALT key
-			 * when changing the TOP will also update BOTTOM.
-			 */
-			// @ts-expect-error - TODO: event.altKey is only present when the change event was
-			// triggered by a keyboard event. Should this feature be implemented differently so
-			// it also works with drag events?
-			if ( event.altKey ) {
-				switch ( side ) {
-					case 'top':
-						nextValues.bottom = nextValue;
-						break;
-					case 'bottom':
-						nextValues.top = nextValue;
-						break;
-					case 'left':
-						nextValues.right = nextValue;
-						break;
-					case 'right':
-						nextValues.left = nextValue;
-						break;
-				}
+		const isNumeric = next !== undefined && ! isNaN( parseFloat( next ) );
+		const nextValue = isNumeric ? next : undefined;
+
+		nextValues[ side ] = nextValue;
+
+		/**
+		 * Supports changing pair sides. For example, holding the ALT key
+		 * when changing the TOP will also update BOTTOM.
+		 */
+		// @ts-expect-error - TODO: event.altKey is only present when the change event was
+		// triggered by a keyboard event. Should this feature be implemented differently so
+		// it also works with drag events?
+		if ( extra?.event.altKey ) {
+			switch ( side ) {
+				case 'top':
+					nextValues.bottom = nextValue;
+					break;
+				case 'bottom':
+					nextValues.top = nextValue;
+					break;
+				case 'left':
+					nextValues.right = nextValue;
+					break;
+				case 'right':
+					nextValues.left = nextValue;
+					break;
 			}
+		}
 
-			handleOnChange( nextValues );
-		};
+		handleOnChange( nextValues );
+	};
 
 	const createHandleOnUnitChange =
 		( side: keyof BoxControlValue ) => ( next?: string ) => {
@@ -132,7 +125,9 @@ export default function BoxInputControls( {
 							value={ [ parsedQuantity, computedUnit ].join(
 								''
 							) }
-							onChange={ createHandleOnChange( side ) }
+							onChange={ ( nextValue, extra ) =>
+								handleOnValueChange( side, nextValue, extra )
+							}
 							onUnitChange={ createHandleOnUnitChange( side ) }
 							onFocus={ createHandleOnFocus( side ) }
 							onHoverOn={ createHandleOnHoverOn( side ) }
@@ -146,9 +141,11 @@ export default function BoxInputControls( {
 							__nextHasNoMarginBottom
 							hideLabelFromVision
 							onChange={ ( newValue ) => {
-								sliderOnChange(
+								handleOnValueChange(
 									side,
-									[ newValue, computedUnit ].join( '' )
+									newValue !== undefined
+										? [ newValue, computedUnit ].join( '' )
+										: undefined
 								);
 							} }
 							min={ 0 }

```
</details>

The main benefit is that we don't duplicate the "on change" logic, which should also help making sure we don't introduce unexpected behaviours

Reply:
Thank you for clarifying and for the diff!